File: apps/desktop/src/renderer/regions.ts
/* A region groups nodes serving one location. Derived from the node id's
   numeric suffix until the hub sends an explicit region key. */

export interface Region {
  key: string;
  name: string;
  country: string;
  nodes: ServerInfo[];
}

const SUFFIX = /-(\d+)$/;

/** `eu-central-1` -> `eu-central`; ids without a numeric suffix stand alone. */
export function regionKeyOf(server: ServerInfo): string {
  return server.id.replace(SUFFIX, "");
}

/** Groups servers into regions, preserving the order they first appear in. */
export function groupRegions(servers: readonly ServerInfo[]): Region[] {
  const byKey = new Map<string, Region>();
  for (const server of servers) {
    const key = regionKeyOf(server);
    const existing = byKey.get(key);
    if (existing) {
      existing.nodes = [...existing.nodes, server];
      continue;
    }
    byKey.set(key, { key, name: server.name || key, country: server.country || "", nodes: [server] });
  }
  return [...byKey.values()];
}

/** Lowest-load node in the region. Nodes with no load report sort last. */
export function pickNode(region: Region): ServerInfo {
  return region.nodes.reduce((best, node) => (loadOf(node) < loadOf(best) ? node : best), region.nodes[0]);
}

/** Missing load is treated as fully loaded so a reporting node always wins. */
export function loadOf(server: ServerInfo): number {
  return typeof server.load === "number" && Number.isFinite(server.load) ? server.load : 100;
}

/** The load actually offered by a region — that of the node we would pick. */
export function regionLoad(region: Region): number | null {
  const node = pickNode(region);
  return typeof node.load === "number" && Number.isFinite(node.load) ? node.load : null;
}

export function findRegion(regions: readonly Region[], key: string): Region | undefined {
  return regions.find((region) => region.key === key);
}

export function regionOfServer(regions: readonly Region[], serverId: string): Region | undefined {
  return regions.find((region) => region.nodes.some((node) => node.id === serverId));
}

/** Selected node, then its siblings by load, then every later region in hub order. */
export function buildServerRetryOrder(servers: readonly ServerInfo[], initialServerId: string): string[] {
  const regions = groupRegions(servers);
  const initialRegion = regionOfServer(regions, initialServerId);
  if (!initialRegion) return [initialServerId];

  const byLoad = (nodes: readonly ServerInfo[]): ServerInfo[] =>
    [...nodes].sort((a, b) => loadOf(a) - loadOf(b));
  const sameRegion = byLoad(initialRegion.nodes)
    .filter((node) => node.id !== initialServerId)
    .map((node) => node.id);
  const initialRegionIndex = regions.indexOf(initialRegion);
  const laterRegions = [...regions.slice(initialRegionIndex + 1), ...regions.slice(0, initialRegionIndex)]
    .flatMap((region) => byLoad(region.nodes).map((node) => node.id));

  return [initialServerId, ...sameRegion, ...laterRegions];
}

/** Most-recent-first, then everything else in hub order. */
export function orderByRecent(regions: readonly Region[], recent: readonly string[]): Region[] {
  const ranked = recent
    .map((key) => regions.find((region) => region.key === key))
    .filter((region): region is Region => region !== undefined);
  const rest = regions.filter((region) => !ranked.includes(region));
  return [...ranked, ...rest];
}

/** Moves `key` to the front, dropping any earlier entry and capping the list. */
export function promoteRecent(recent: readonly string[], key: string, limit = 8): string[] {
  return [key, ...recent.filter((entry) => entry !== key)].slice(0, limit);
}

File: apps/android/core/src/main/kotlin/org/pangeavpn/core/Regions.kt
package org.pangeavpn.core

import kotlin.random.Random
import org.pangeavpn.core.model.Server

/** A region groups the nodes serving one location. Ports
 *  apps/desktop/src/renderer/{regions,serverPick}.ts. */
data class Region(
    val key: String,
    val name: String,
    val country: String,
    val nodes: List<Server>,
)

private val NODE_SUFFIX = Regex("-(\\d+)$")

/** Servers at or below this load are preferred by a random pick. */
const val PREFERRED_MAX_LOAD = 75

/** `eu-central-1` becomes `eu-central`; ids without a suffix stand alone. */
fun regionKeyOf(server: Server): String = server.id.replace(NODE_SUFFIX, "")

/** Groups servers into regions, preserving first-seen order. */
fun groupRegions(servers: List<Server>): List<Region> {
    val byKey = LinkedHashMap<String, Region>()
    for (server in servers) {
        val key = regionKeyOf(server)
        val existing = byKey[key]
        byKey[key] = existing?.copy(nodes = existing.nodes + server)
            ?: Region(
                key = key,
                name = server.name.ifEmpty { key },
                country = server.country,
                nodes = listOf(server),
            )
    }
    return byKey.values.toList()
}

/** Missing load counts as fully loaded, so a node that reports always wins. */
fun loadOf(server: Server): Int = server.load ?: 100

/** Lowest-load node in the region. */
fun pickNode(region: Region): Server =
    region.nodes.minByOrNull { loadOf(it) } ?: region.nodes.first()

/** The load a region actually offers: that of the node we would pick. */
fun regionLoad(region: Region): Int? = pickNode(region).load

fun regionOfServer(regions: List<Region>, serverId: String): Region? =
    regions.firstOrNull { region -> region.nodes.any { it.id == serverId } }

/** How many regions get a row of their own before the rest go behind "All regions". */
const val REGION_SLOT_COUNT = 2

/**
 * The regions shown up front. Whatever is selected is always one of them, even
 * when it is not among the lightest — a picked region that vanished from view
 * would read as though the pick had been lost.
 */
fun regionSlots(regions: List<Region>, selectedKey: String?, slots: Int = REGION_SLOT_COUNT): List<Region> {
    val head = regions.take(slots)
    if (selectedKey == null || head.any { it.key == selectedKey }) return head
    val selected = regions.firstOrNull { it.key == selectedKey } ?: return head
    return (listOf(selected) + regions.filter { it.key != selectedKey }).take(slots)
}

/**
 * Uniform pick from the lightly loaded servers, falling back to all of them
 * when every server is congested or none reports load. Deliberately not
 * load-proportional, so clients do not herd onto one node.
 */
fun pickRandomServer(servers: List<Server>, random: Random = Random.Default): Server? {
    if (servers.isEmpty()) return null
    val light = servers.filter { server -> server.load?.let { it <= PREFERRED_MAX_LOAD } == true }
    val pool = light.ifEmpty { servers }
    return pool[random.nextInt(pool.size)]
}

/**
 * Which server the picker shows: the in-session pick if still listed, else the
 * last connected one if still listed, else nothing — so the picker reads
 * "Select server" and Connect rolls a random one.
 */
fun resolveSelection(visible: List<Server>, previousValue: String, lastServerId: String?): String {
    if (visible.any { it.id == previousValue }) return previousValue
    if (lastServerId != null && visible.any { it.id == lastServerId }) return lastServerId
    return ""
}
